torchx/c_src/torchx.cpp

```diff
@@ -0,0 +1,420 @@
+  SHAPE_PARAM(1, shape);
+
+  TENSOR(at::reshape(*t, shape));
+}
+
+NIF(to_type)
+{
+  TENSOR_PARAM(0, t);
+  TYPE_PARAM(1, type);
+
+  TENSOR(t->toType(type));
+}
+
+NIF(squeeze)
+{
+  TENSOR_PARAM(0, t);
+
+  if (argc == 2)
+  {
+    PARAM(1, int64_t, dim);
+    TENSOR(at::squeeze(*t, dim));
+  }
+  else
+    TENSOR(at::squeeze(*t));
+}
+
+NIF(broadcast_to)
+{
+  TENSOR_PARAM(0, t);
+  SHAPE_PARAM(1, shape);
+
+  TENSOR(at::broadcast_to(*t, shape).clone());
+}
+
+NIF(transpose)
+{
+  TENSOR_PARAM(0, t);
+  PARAM(1, int64_t, dim0);
+  PARAM(2, int64_t, dim1);
+
+  TENSOR(at::transpose(*t, dim0, dim1));
+}
+
+NIF(ones)
+{
+  SHAPE_PARAM(0, shape);
+  TYPE_PARAM(1, type);
+
+  TENSOR(at::ones(shape, type));
+}
+
+NIF(eye)
+{
+  PARAM(0, int64_t, size);
+  TYPE_PARAM(1, type);
+
+  TENSOR(at::eye(size, type));
+}
+
+NIF(add)
+{
+  TENSOR_PARAM(0, a);
+  TENSOR_PARAM(1, b);
+
+  if (b == NULL)
+  {
+    PARAM(1, double, scalar);
+    TENSOR(*a + scalar);
+  }
+  else
+    TENSOR(*a + *b);
+}
+
+NIF(dot)
+{
+  TENSOR_PARAM(0, a);
+  TENSOR_PARAM(1, b);
+
+  TENSOR(at::matmul(*a, *b));
+}
+
+NIF(cholesky)
+{
+  TENSOR_PARAM(0, t);
+  bool upper = false;
+
+  if (argc == 2)
+  {
+    GET(1, upper);
+  }
+
+  TENSOR(at::cholesky(*t, upper));
+}
+
+NIF(qr)
+{
+  TENSOR_PARAM(0, t);
+  bool reduced = true;
+
+  if (argc == 2)
+  {
+    GET(1, reduced);
+  }
+
+  TENSOR_TUPLE(at::qr(*t, reduced));
+}
+
+void free_tensor(ErlNifEnv *env, void *obj)
+{
+  std::cout << "Deleting: " << obj << std::endl;
+}
+
+static int
+open_resource_type(ErlNifEnv *env)
+{
+  const char *name = "Tensor";
+  ErlNifResourceFlags flags = (ErlNifResourceFlags)(ERL_NIF_RT_CREATE | ERL_NIF_RT_TAKEOVER);
+
+  TENSOR_TYPE = enif_open_resource_type(env, NULL, name, free_tensor, flags, NULL);
+  if (TENSOR_TYPE == NULL)
+    return -1;
+  return 0;
+}
+
+int upgrade(ErlNifEnv *env, void **priv_data, void **old_priv_data, ERL_NIF_TERM load_info)
+{
+  // Silence "unused var" warnings.
+  (void)(env);
+  (void)(priv_data);
+  (void)(old_priv_data);
+  (void)(load_info);
+
+  return 0;
+}
+
+int load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info)
+{
+  if (open_resource_type(env) == -1)
+    return -1;
+
+  // Silence "unused var" warnings.
+  (void)(priv_data);
+  (void)(load_info);
+
+  return 0;
+}
+
+#define F(NAME, ARITY)    \
+  {                       \
+#NAME, ARITY, NAME, 0 \
+  }
+
+static ErlNifFunc nif_functions[] = {
+    F(randint, 4),
+    F(rand, 4),
+    F(normal, 3),
+    F(arange, 4),
+    F(arange, 5),
+    F(from_blob, 3),
+    F(to_blob, 1),
+    F(to_blob, 2),
+    F(type, 1),
+    F(device, 1),
+    F(nbytes, 1),
+    F(scalar_tensor, 2),
+    F(delete_tensor, 1),
+    F(ones, 1),
+    F(eye, 2),
+    F(reshape, 2),
+    F(split, 2),
+    F(to_type, 2),
+    F(squeeze, 2),
+    F(squeeze, 1),
+    F(broadcast_to, 2),
+    F(transpose, 3),
+    F(add, 2),
+    F(dot, 2),
+    F(cholesky, 1),
+    F(cholesky, 2),
+    F(qr, 1),
+    F(qr, 2)};
```

Comment:
We should split those functions. Most of them should be tagged as CPU bound Dirty NIFs, except for a handful (`type`, `device`, `nbytes`).

In the future, we want to have both CPU and IO dirty NIFs (the IO dirty ones for GPU stuff) - and we will need a way to control which one we want to call.

Reply:
I added two definitions for each fun — the default is CPU dirty and _io version, which is IO dirty.